contrib/recipes/key.go

```diff
@@ -0,0 +1,190 @@
+}
+
+func NewKV(client *EtcdClient, key, val string, leaseID lease.LeaseID) (*RemoteKey, error) {
+	rev, err := putNewKV(client, key, val, leaseID)
+	if err != nil {
+		return nil, err
+	}
+	return &RemoteKey{client, key, rev, val}, nil
+}
+
+func GetRemoteKey(client *EtcdClient, key string) (*RemoteKey, error) {
+	resp, err := client.KV.Range(
+		context.TODO(),
+		&pb.RangeRequest{Key: []byte(key)},
+	)
+	if err != nil {
+		return nil, err
+	}
+	rev := resp.Header.Revision
+	val := ""
+	if len(resp.Kvs) > 0 {
+		rev = resp.Kvs[0].ModRevision
+		val = string(resp.Kvs[0].Value)
+	}
+	return &RemoteKey{
+		client: client,
+		key:    key,
+		rev:    rev,
+		val:    val}, nil
+}
+
+func NewUniqueKey(client *EtcdClient, prefix string) (*RemoteKey, error) {
+	return NewUniqueKV(client, prefix, "", 0)
+}
+
+func NewUniqueKV(client *EtcdClient, prefix string, val string, leaseID lease.LeaseID) (*RemoteKey, error) {
+	for {
+		newKey := fmt.Sprintf("%s/%v", prefix, time.Now().UnixNano())
+		rev, err := putNewKV(client, newKey, val, 0)
+		if err == nil {
+			return &RemoteKey{client, newKey, rev, val}, nil
+		}
+		if err != ErrKeyExists {
+			return nil, err
+		}
+	}
+}
+
+// putNewKV attempts to create the given key, only succeeding if the key did
+// not yet exist.
+func putNewKV(ec *EtcdClient, key, val string, leaseID lease.LeaseID) (int64, error) {
+	cmp := &pb.Compare{
+		Result: pb.Compare_EQUAL,
+		Target: pb.Compare_VERSION,
+		Key:    []byte(key)}
+	req := &pb.RequestUnion{
+		RequestPut: &pb.PutRequest{
+			Key:   []byte(key),
+			Value: []byte(val),
+			Lease: int64(leaseID)}}
+
+	txnresp, err := ec.KV.Txn(
+		context.TODO(),
+		&pb.TxnRequest{[]*pb.Compare{cmp}, []*pb.RequestUnion{req}, nil})
+	if err != nil {
+		return 0, err
+	}
+	if txnresp.Succeeded == false {
+		return 0, ErrKeyExists
+	}
+	return txnresp.Header.Revision, nil
+}
+
+// NewSequentialKV allocates a new sequential key-value pair at <prefix>/nnnnn
+func NewSequentialKV(client *EtcdClient, prefix, val string) (*RemoteKey, error) {
+	return newSequentialKV(client, prefix, val, 0)
+}
+
+// newSequentialKV allocates a new sequential key <prefix>/nnnnn with a given
+// value and lease.  Note: a bookkeeping node __<prefix> is also allocated.
+func newSequentialKV(client *EtcdClient, prefix, val string, leaseID lease.LeaseID) (*RemoteKey, error) {
+	resp, err := NewRange(client, prefix).LastKey()
+	if err != nil {
+		return nil, err
+	}
+
+	// add 1 to last key, if any
+	newSeqNum := 0
+	if len(resp.Kvs) != 0 {
+		fields := strings.Split(string(resp.Kvs[0].Key), "/")
+		_, err := fmt.Sscanf(fields[len(fields)-1], "%d", &newSeqNum)
+		if err != nil {
+			return nil, err
+		}
+		newSeqNum++
+	}
+	newKey := fmt.Sprintf("%s/%016d", prefix, newSeqNum)
+
+	// base prefix key must be current (i.e., <=) with the server update;
+	// the base key is important to avoid the following:
+	// N1: LastKey() == 1, start txn.
+	// N2: New Key 2, New Key 3, Delete Key 2
```

Comment:
There's only one consumer of sequential keys: priority queues. The removal happens when dequeuing.

The contention is gross but it's only for priority queues (which mitigates some contention since the base keys are per priority); I don't think server support for sequential keys is worthwhile just yet. That said, I suspect there's a better way to get sequential keys with the existing v3 api.